# Classification performance of the keyword search algorithm

The data and code here produces the in-sample and out-of-sample classification performance statistics.

I use keyword search to find forward-looking sentences. The keyword list for forward-looking is uploaded here. Besides, I determine that if a sentence contains words of years for the coming ten years, it is also forward-looking. The word list is adapted from Muslu et al. [2015].

Among the forward-looking statements, I use a time frame keyword list to identify those with time frames. The keyword list for time frames is uploaded here. The time frame keyword list is adapted from Brochet, Loumioti and Serafeim [2015].

In [1]:
import pandas as pd
gold_standard = pd.read_excel("Gold Standard Corpus.xlsx")
gold_standard[['file_name', 'last_update', 'section', 'context', 
               'speaker_number', 'future_year',
               'speaker_text', 'fl', 'qt', 'fl_qt', 'fl_cal',
               'qt_cal', 'fl_qt_cal', 'training sample', 'test sample']]

file_name                last_update  section context  speaker_number  \
0     12614899_T  2019-08-01 19:19:18+10:00        1      qa              23   
1     12614899_T  2019-08-01 19:19:18+10:00        1      qa              23   
2     12614899_T  2019-08-01 19:19:18+10:00        1      qa              23   
3       702568_T  2003-01-30 16:25:51+11:00        1      qa             109   
4       702568_T  2003-01-30 16:25:51+11:00        1      qa             109   
...          ...                        ...      ...     ...             ...   
1283   2723598_T  2010-02-17 13:53:38+11:00        1    pres               5   
1284   2723598_T  2010-02-17 13:53:38+11:00        1    pres               5   
1285    739440_T  2003-05-07 12:37:12+10:00        1      qa             154   
1286    739440_T  2003-05-07 12:37:12+10:00        1      qa             154   
1287    739440_T  2003-05-07 12:37:12+10:00        1      qa             154   

      future_year                                       speaker_text  fl  qt  \
0            2019                                              Okay.   0   0   
1            2019  And lastly from me, can you kind of give us a ...   0   0   
2            2019     What's the share of wallet for Knoll products?   0   0   
3            2003                                              Okay.   0   0   
4            2003           So I don't think anybody said much more.   0   0   
...           ...                                                ...  ..  ..   
1283         2010  With that, let's open the call to your questions.   0   0   
1284         2010           Yvette, please review the Q&A procedure.   0   0   
1285         2003  This is a follow up question from Chris Joseph...   0   0   
1286         2003                                  from [inaudible].   0   0   
1287         2003                                   Please go ahead.   0   0   

      fl_qt  fl_cal  qt_cal  fl_qt_cal  training sample  test sample  
0         0       0       0          0                0            0  
1         0       0       0          0                1            0  
2         0       0       0          0                1            0  
3         0       0       0          0                1            0  
4         0       0       0          0                1            0  
...     ...     ...     ...        ...              ...          ...  
1283      0       0       0          0                0            1  
1284      0       0       0          0                1            0  
1285      0       0       0          0                1            0  
1286      0       0       0          0                1            0  
1287      0       0       0          0                0            1  

[1288 rows x 15 columns]

The underlying text associated with the gold standard is found in [`Gold Standard Corpus.xlsx`](https://github.com/yiyangw2/time_frame_gold_corpus/blob/main/Gold%20Standard%20Corpus.xlsx). It is composed of 200 randomly picked up paragraphs of the management (1288 sentences). As you can see, each sentence has identifiers for where it comes from. And values for `fl_cal`,`qt_cal` and `fl_qt_cal`, reflecting whether the algorithm determines that whether the sentence is forward-looking, whether it contains time frames and whether it is a forward-looking statement with time frames. Similarly, each observation has values for `fl`,`qt` and `fl_qt`, which is whether I manually determine that whether the sentence is forward-looking, whether it contains time frames and whether it is a forward-looking statement with time frames. The underlying earnings calls' scripts are from the Melbourne Centre for Corporate Governance and Regulation (MCCGR) database. 

The following function applies the identification functions to each sentence and shows how `fl_cal`, `qt_cal` and `fl_qt_cal` in `gold_standard` are produced.

In [2]:
from time_frame import is_forward_looking, is_qt_term, print_stats

In [3]:
gold_standard['fl_cal'] = gold_standard.apply(lambda x: is_forward_looking(x['speaker_text'].lower(), x['future_year']), axis=1)
gold_standard['qt_cal'] = gold_standard.apply(lambda x: is_qt_term(x['speaker_text'].lower()), axis=1)
gold_standard["fl_qt_cal"] = gold_standard.fl_cal & gold_standard.qt_cal

The following function generates statistics to report the performance of the algorithm on identifying forward-looking statements and forward-looking statements with time frames.

In [4]:
training_sample = gold_standard[gold_standard['training sample']==1]
test_sample = gold_standard[gold_standard['training sample']==0]

In [5]:
print_stats(training_sample, type = 'fl')

Accuracy 95.23%
Precision 82.09%
True positive rate 85.94%


In [6]:
print_stats(test_sample, type = 'fl')

Accuracy 95.82%
Precision 83.61%
True positive rate 87.93%


In [7]:
print_stats(training_sample, type = 'fl_qt')

Accuracy 96.37%
Precision 88.64%
True positive rate 59.09%


In [8]:
print_stats(test_sample, type = 'fl_qt')

Accuracy 96.56%
Precision 75.00%
True positive rate 69.23%
